pkg/netdevice/vdpa.go

```diff
@@ -0,0 +1,139 @@
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+const (
+	vdpaTypeVirtio = "virtio"
+	vdpaTypeVhost  = "vhost"
+)
+
+/*vdpaTypeToDriver translates vdpaTypes (as specified in the netDevice selectors)
+  to vdpa bus drivers*/
+var vdpaTypeToDriverMap = map[string]string{
+	vdpaTypeVirtio: vdpa.VirtioVdpaDriver,
+	vdpaTypeVhost:  vdpa.VhostVdpaDriver,
+}
+
+func vdpaTypeToDriver(vType string) (driver string, exists bool) {
+	driver, exists = vdpaTypeToDriverMap[vType]
+	return
+}
+
+func vdpaDriverToType(vDriver string) (string, bool) {
+	for t, d := range vdpaTypeToDriverMap {
+		if d == vDriver {
+			return t, true
+		}
+	}
+	return "", false
+}
+
+type vdpaInfo struct {
+	device vdpa.VdpaDevice
+}
+
+// GetVdpaDevice returns a VdpaDevice
+func GetVdpaDevice(pciAddr string) vdpa.VdpaDevice {
+	vdpaDev, err := vdpa.GetVdpaDeviceByPci(pciAddr)
+	if err != nil {
+		glog.Infof("%s - No vDPA device found: %s", pciAddr, err)
+		return nil
+	}
+	return vdpaDev
+}
+
+// vdpaInfoProvider is a DeviceInfoProvider that handles the API information of vdpa-capable devices
+type vdpaInfoProvider struct {
+	dev      vdpa.VdpaDevice
+	vdpaType string
+}
+
+// NewVdpaInfoProvider returns a new InfoProvider associated with the given VDPAInfo
+func NewVdpaInfoProvider(vdpaType string, vdpaDev vdpa.VdpaDevice) types.DeviceInfoProvider {
+	vdpaInfoProvider := &vdpaInfoProvider{
+		vdpaType: vdpaType,
+		dev:      vdpaDev,
+	}
+	if err := vdpaInfoProvider.Error(); err != nil {
+		glog.Warning(err)
+	}
+	return vdpaInfoProvider
+}
+
+// GetDeviceSpecs returns the DeviceSpec slice
+func (vip *vdpaInfoProvider) GetDeviceSpecs() []*pluginapi.DeviceSpec {
+	if err := vip.Error(); err != nil {
+		glog.Errorf("GetDeviceSpecs(): vDPA is required in the configuration but device does not have a healthy vdpa device: %s",
+			err)
+		return nil
+	}
+	devSpecs := make([]*pluginapi.DeviceSpec, 0)
+
+	// DeviceSpecs only required for vhost vdpa type as the
+	if vip.vdpaType == vdpaTypeVhost {
+		devSpecs = append(devSpecs, &pluginapi.DeviceSpec{
+			HostPath:      vip.dev.GetPath(),
+			ContainerPath: vip.dev.GetPath(),
+			Permissions:   "mrw",
+		})
+	}
+	return devSpecs
+}
+
+// GetEnvVal returns the environment variable value
+/*
+   This Info Provider does not assume there is any other Info Provider present.
+   Thus, it sets the Env Value as we don't support multiple values or their
+   concatenation for now.
+*/
+func (vip *vdpaInfoProvider) GetEnvVal() string {
+	if vip.Error() == nil {
+		return fmt.Sprintf("%s:%s", vip.dev.GetDriver(), vip.dev.GetPath())
+	}
+	return ""
+}
+
+// GetMounts returns the mount points (none for this InfoProvider)
+func (vip *vdpaInfoProvider) GetMounts() []*pluginapi.Mount {
+	mounts := make([]*pluginapi.Mount, 0)
+	return mounts
+}
+
+// Error returns an error if the vdpaInfoProvider is not valid
+func (vip *vdpaInfoProvider) Error() error {
```

Comment:
perhaps this should be internal method ? as code outside this file will only deal with DeviceInfoProvider interface

Reply:
Agree. Reworded into isHealthy and made private